autoload/go/package.vim

```diff
@@ -144,12 +144,31 @@ function! go#package#FromPath(arg) abort
   let l:cd = exists('*haslocaldir') && haslocaldir() ? 'lcd' : 'cd'
   let l:dir = getcwd()
 
-  let l:path = a:arg
+  let l:path = fnamemodify(a:arg, ':p')
   if !isdirectory(l:path)
     let l:path = fnamemodify(l:path, ':h')
   endif
 
   execute l:cd fnameescape(l:path)
+  if glob("*.go") == ""
+    " There's no Go code in this directory. We might be in a module directory
+    " which doesn't have any code at this level.
+    let l:module = s:module()
+    if !empty(l:module)
+      let l:importpath = l:module.path
+
+      " Build up a fake import path by combining the module path and the path
+      " to the target directory (relative to the module diretory).
+      execute l:cd fnameescape(l:module.dir)
+      let l:relativepath = fnamemodify(l:path, ':.')
+      if l:relativepath != l:path
```

Comment:
I think you meant `l:path` is always an absolute path now (`:p` converts
it to a full path).

`l:relativepath` will be the same as `l:path` if `fnamemodify(.., ':.')`
cannot make `l:path` relative to `module.dir`. Although now that I think
about it, that'll happen only if `l:path == module.dir`. I guess that's
moot, though, because your solution to this problem is much cleaner.